### PR TITLE
WT3 surface discharge ZO model

### PR DESCRIPTION
**Pull request by tarnold17**
## Summary/Motivation:
Migrate surface discharge ZO model from WT3

## Changes proposed in this PR:
- Add surface discharge ZO model and unit tests
- Update YAML

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/core/zero_order_costing.py [hunk 1]
```diff
@@ -939,6 +940,49 @@ def cost_storage_tank(blk):
...
+            initialize=1,
+            units=blk.config.flowsheet_costing_block.base_currency,
+            bounds=(0, None),
+            doc="Capital cost of unit operation")
+
+        Q = pyo.units.convert(
+            blk.unit_model.properties[t0].flow_vol,
+            to_units=pyo.units.m ** 3 / pyo.units.hour)
+        expr = pyo.units.convert(
+            A * (Q / ref_state) ** B
```
Comment (andrewlee94) [inline watertap/core/zero_order_costing.py]: In this case, what you really want ot do is `pyo.units.convert(flow_vol/ref_state, to_units=pyo.units.dimensionless)`.

This allows for future cases where the units of `ref_state` are not known.

File: watertap/data/techno_economic/surface_discharge.yaml [hunk 2]
```diff
@@ -1,9 +1,49 @@
...
+    units: miles
+emwd:
+  capital_cost:
+    basis: flow_vol
+    cost_factor: None
+    reference_state:
+      value: 10417
+      units: m^3/hr
+    capital_a_parameter:
+      value: 35e6 # TODO does this need to be scaled to account for $ vs MM$ ?
```
Comment (andrewlee94) [inline watertap/data/techno_economic/surface_discharge.yaml]: Most $ values in the WT3 code were for millions of dollars ,so we generally have to multiply by 1e6 in our YAML files (alternatively, you could set the cost units as `MUSD_2020`, but we have generally just used 1e6).

File: watertap/unit_models/zero_order/surface_discharge_zo.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                                 doc='Piping distance')
+
+        self.pipe_diameter = Var(self.flowsheet().config.time,
+                                 units=pyunits.inches,
+                                 doc='Pipe diameter')
+
+        self._fixed_perf_vars.append(self.pipe_distance)
+        self._fixed_perf_vars.append(self.pipe_diameter)
+
+        self._perf_var_dict["Pipe Distance (miles)"] = self.pipe_distance
```
Comment (andrewlee94) [inline watertap/unit_models/zero_order/surface_discharge_zo.py]: We generally do not put units in here, as we will (eventually) report units with the values in the report output. and thus these hard-coded units might not be correct.

File: watertap/unit_models/zero_order/tests/test_surface_discharge.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+class TestSurfaceDischargeZO:
+    @pytest.fixture(scope="class")
+    def model(self):
+        m = ConcreteModel()
+        m.db = Database()
+
+        m.fs = FlowsheetBlock(default={"dynamic": False})
+        m.fs.params = WaterParameterBlock(
+                default={"solute_list": ["toc", "nitrate", "sulfate", "bar", "crux"]})
```
Comment (andrewlee94) [inline watertap/unit_models/zero_order/tests/test_surface_discharge.py]: What is `crux` in this case, and why is it needed?

**Review by andrewlee94 — CHANGES_REQUESTED** [review]
A couple of minor comments, but otherwise looks good. Well done for your first model.

File: watertap/core/zero_order_costing.py (hunk 1 above)
Comment (tarnold17) [reply watertap/core/zero_order_costing.py]: I'm not sure that I precisely follow your suggestion here - are you saying that I need to wrap `(Q / ref_state)` in the call to `pyo.units.convert()` that you suggested? I.e., I should have two calls to `pyo.units.convert()`, one nested inside the other?

File: watertap/data/techno_economic/surface_discharge.yaml (hunk 2 above)
Comment (tarnold17) [reply watertap/data/techno_economic/surface_discharge.yaml]: Is there an action required here? The A parameter that I found in WT3 was 35 [(see here)](https://github.com/NREL/WaterTAP3/blob/master/watertap3/watertap3/wt_units/surface_discharge.py) so I have multiplied by 1e6.

File: watertap/unit_models/zero_order/surface_discharge_zo.py (hunk 3 above)
Comment (tarnold17) [reply watertap/unit_models/zero_order/surface_discharge_zo.py]: So these strings should simply read `"Pipe Distance"` and `"Pipe Diameter"`? I can change that. This is something that I just copied and pasted from the well field ZO model.

File: watertap/unit_models/zero_order/tests/test_surface_discharge.py (hunk 4 above)
Comment (tarnold17) [reply watertap/unit_models/zero_order/tests/test_surface_discharge.py]: @kurbansitterley told me that it is just meant to be a dummy component. This is another instance of copy/paste from the well field ZO model. I can remove all references to "crux" if you'd like.

File: watertap/core/zero_order_costing.py (hunk 1 above)
Comment (andrewlee94) [reply watertap/core/zero_order_costing.py]: Yes, you can nest convert statements as needed. So:

```
expr = pyo.units.convert(
            A * pyo.units.convert(Q / ref_state, to_units=pyo.units.dimensionless)**B,
            to_units=base_currency)
```

File: watertap/data/techno_economic/surface_discharge.yaml (hunk 2 above)
Comment (andrewlee94) [reply watertap/data/techno_economic/surface_discharge.yaml]: No, I don't think so. I just wanted to clarify,

File: watertap/unit_models/zero_order/surface_discharge_zo.py (hunk 3 above)
Comment (andrewlee94) [reply watertap/unit_models/zero_order/surface_discharge_zo.py]: I think that is better. It is on the list of thing to do in IDAES to add units automatically to these reports.

File: watertap/unit_models/zero_order/tests/test_surface_discharge.py (hunk 4 above)
Comment (andrewlee94) [reply watertap/unit_models/zero_order/tests/test_surface_discharge.py]: You already have `"bar"` for a dummy component in this case however.

Comment (tarnold17) [reply watertap/unit_models/zero_order/tests/test_surface_discharge.py]: True. I will remove "crux" on my next update.

**Review by luohezhiming — DISMISSED** [review]
Looks good to me.

File: watertap/core/zero_order_costing.py (hunk 1 above)
Comment (tarnold17) [reply watertap/core/zero_order_costing.py]: I addressed this on my most recent update, so I am marking resolved here.

File: watertap/unit_models/zero_order/surface_discharge_zo.py (hunk 3 above)
Comment (tarnold17) [reply watertap/unit_models/zero_order/surface_discharge_zo.py]: I addressed this on my most recent update, so I am marking resolved here.

File: watertap/unit_models/zero_order/tests/test_surface_discharge.py (hunk 4 above)
Comment (tarnold17) [reply watertap/unit_models/zero_order/tests/test_surface_discharge.py]: I addressed this on my most recent update, so I am marking resolved here.

File: watertap/core/zero_order_costing.py [hunk 5]
```diff
@@ -972,7 +972,7 @@ def cost_surface_discharge(blk):
             blk.unit_model.properties[t0].flow_vol,
             to_units=pyo.units.m ** 3 / pyo.units.hour)
         expr = pyo.units.convert(
-            A * (Q / ref_state) ** B
+            A * pyo.units.convert(Q / ref_state, to_units=pyo.units.dimensionless) ** B
```
Comment (andrewlee94) [inline watertap/core/zero_order_costing.py]: By doing this, you no longer need the unit conversion for `Q` above - you can just use `blk.unit_model.properties[t0].flow_vol` here.

**Review by andrewlee94 — COMMENTED** [review]
One more small comment to fix.

File: watertap/core/zero_order_costing.py [hunk 6]
```diff
@@ -939,6 +940,46 @@ def cost_storage_tank(blk):
...
+        # Add cost variable and constraint
+        blk.capital_cost = pyo.Var(
+            initialize=1,
+            units=blk.config.flowsheet_costing_block.base_currency,
+            bounds=(0, None),
+            doc="Capital cost of unit operation")
+
+        expr = pyo.units.convert(
+            A * pyo.units.convert(blk.unit_model.properties[t0].flow_vol / ref_state, to_units=pyo.units.dimensionless) ** B
+            + (pipe_cost_basis * blk.unit_model.pipe_distance[t0] * blk.unit_model.pipe_diameter[t0]),
```
Comment (andrewlee94) [inline watertap/core/zero_order_costing.py]: You should probably separate the unit conversion to be around each part of the sum here. I.e:

```
expr = (
    pyo.units.convert(
        A * pyo.units.convert(blk.unit_model.properties[t0].flow_vol / ref_state, to_units=pyo.units.dimensionless) ** B,
        to_units=blk.config.flowsheet_costing_block.base_currency) +
    pyo.units.convert(
        pipe_cost_basis * blk.unit_model.pipe_distance[t0] * blk.unit_model.pipe_diameter[t0]),
        to_units=blk.config.flowsheet_costing_block.base_currency))
```

The reason for this is that Pyomo will object if the two parts of the sum do not have same units before conversion. This way, we force each part of the sum to have the same units before doing the addition.

File: watertap/data/techno_economic/surface_discharge.yaml (hunk 2 above)
Comment (andrewlee94) [inline watertap/data/techno_economic/surface_discharge.yaml]: You should probably remove this comment no that this is resolved.

**Review by andrewlee94 — COMMENTED** [review]
Two more small things to fix.

**Review by andrewlee94 — APPROVED** [review]

**Review by luohezhiming — APPROVED** [review]

**Pull request merged**